sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/SpiUtil.java

```diff
@@ -9,39 +9,100 @@
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Optional;
 import java.util.ServiceLoader;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.function.BiFunction;
 import java.util.function.Function;
+import java.util.function.Supplier;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+import javax.annotation.Nullable;
 
 final class SpiUtil {
 
   private static final Logger logger = Logger.getLogger(SpiUtil.class.getName());
 
-  static <T, U> Map<String, T> loadConfigurable(
+  interface ServiceLoaderFinder {
+
+    <S> Iterable<S> load(Class<S> spiClass, ClassLoader classLoader);
+  }
+
+  static <T, U> NamedSpiManager<T> loadConfigurable(
       Class<U> spiClass,
       Collection<String> requestedNames,
       Function<U, String> getName,
       BiFunction<U, ConfigProperties, T> getConfigurable,
       ConfigProperties config,
       ClassLoader serviceClassLoader) {
-    Map<String, T> result = new HashMap<>();
-    for (U provider : ServiceLoader.load(spiClass, serviceClassLoader)) {
+    return loadConfigurable(
+        spiClass,
+        requestedNames,
+        getName,
+        getConfigurable,
+        config,
+        serviceClassLoader,
+        ServiceLoader::load);
+  }
+
+  // VisibleForTesting
+  static <T, U> NamedSpiManager<T> loadConfigurable(
+      Class<U> spiClass,
+      Collection<String> requestedNames,
+      Function<U, String> getName,
+      BiFunction<U, ConfigProperties, T> getConfigurable,
+      ConfigProperties config,
+      ClassLoader serviceClassLoader,
+      ServiceLoaderFinder serviceLoaderFinder) {
+    Map<String, Supplier<T>> nameToProvider = new HashMap<>();
+    for (U provider : serviceLoaderFinder.load(spiClass, serviceClassLoader)) {
       String name = getName.apply(provider);
-      final T configurable;
-      try {
-        configurable = getConfigurable.apply(provider, config);
-      } catch (Throwable t) {
-        Level level = requestedNames.contains(name) ? Level.WARNING : Level.FINE;
-        logger.log(
-            level, "Error initializing " + spiClass.getSimpleName() + " with name " + name, t);
-        continue;
-      }
-      result.put(name, configurable);
+      nameToProvider.put(
+          name,
+          () -> {
+            try {
+              return getConfigurable.apply(provider, config);
+            } catch (Throwable t) {
+              Level level = requestedNames.contains(name) ? Level.WARNING : Level.FINE;
+              logger.log(
+                  level,
+                  "Error initializing " + spiClass.getSimpleName() + " with name " + name,
+                  t);
+              return null;
```

Comment:
If this fails to initialize you probably get a sort of confusing error in some areas i.e. you'll get the above log statement saying it couldn't initialize but then you'll get a subsequent error message saying something along the lines of the provider with <name> wasn't found, since this doesn't differentiate didn't exist and failed to load. Technically because this is now lazily initialized this could probably be changed to not swallow the exception although it's a behavior change.

Reply:
Looking at the change where this was added https://github.com/open-telemetry/opentelemetry-java/pull/3579 it looks like mainly this was due to the eagerness so I think I will move forward with removing the exception catching here and just allow it to propagate.